### ISSUE TITLE
np.roll produces unexpected behavior when shift is a tuple but axis is not

### ISSUE THREAD

**Issue by Sohl-Dickstein** [numpy/numpy#14277]
If the argument `shift` is an array (or similar datatype), but `axis` is None or a scalar, then the behavior of np.roll is undefined by the docs. In practice, np.roll silently does unintuitive things rather than throwing an error. For instance, running
``` python
x = np.array([[1,2],[3,4]])
print('x\n', x)
y = np.roll( x, [-1,1] )
print('y\n', y)
y_expected = np.roll( x, [-1,1], (0,1) )
print('y_expected\n', y_expected)
z = np.roll( x, [1,2,-3] )
print('z\n', z)
```
produces output
> x
>  [[1 2]
>  [3 4]]
> y
>  [[1 2]
>  [3 4]]
> y_expected
>  [[4 3]
>  [2 1]]
> z
>  [[1 2]
>  [3 4]]
Here, `z` is identical to `y` is identical to `x`. However, a naive expectation would be that `y` would produce the same output as `y_expected`, and that `z` would generate an error.

The numpy documentation states that if `shift` is a tuple, then `axis` must be a tuple of the same length. I believe the easiest fix would be to raise an error if `shift` and `axis` have different lengths.

**Comment by seberg** [comment]
Well, if `axis=None` the array is flattened (and then all shifts applied). Otherwise the axis and the shift are broadcast to each other. I would be OK with deprecating this to some degree, but the `None` behaviour of flattening is in some sense well defined. Repeating axis seems also well defined (although in practice probably rarely used). Note that if  give a 2D input without an axis you also do not roll the 0s axis.

**Comment by Sohl-Dickstein** [comment]
I'm OK with flattening, and then shifting, when `axis=None`. The behavior which I think is dangerous is that, when `shift` is a tuple or list, all the shifts are applied to the same axis (of the flattened array). I think if the user is passing in a list of shifts, the expectation will almost never be that the elements in that list will be summed, and then applied along one axis (or equivalently, applied serially along one axis). I had a realization, on understanding this behavior, that I had bugs going back years due to my incorrect assumptions about np.roll.

The documentation explicitly forbids calling np.roll with `shift` as a tuple, but `axis` not a tuple:
> shift : int or tuple of ints
> ... If a tuple, then axis must be a tuple of the same size

So, I think this is a usage pattern which is explicitly forbidden by the documentation, and which will in nearly every case correspond to a code bug. I think it would make a lot of sense to throw an error if `shift` and `axis` have different lengths.

Which is a long explanation -- but I think we may be agreeing with each other! Depending on whether you believe that deprecating to some degree is consistent with throwing an error in this situation.

(I do think a more intuitive default behavior would be something like:
``` python
if axis is None and len(shift) > 1:
    axis = range(len(shift))
```
but I totally understand that silently changing undocumented behavior requires meeting a much higher bar than adding a new error message for undocumented behavior.)

**Comment by PrometheusPi** [comment]
I agree with @Sohl-Dickstein - adding at least clearer documentation or better throwing an error when `axis=None` and the `shift` is a tuple with `len(shift) >1` would be a much better solution.
I just ran into the same *strange* behavior and it took me a while to understand how `roll` operates on multi-dimensional arrays. 

Yes, @seberg everything is correctly stated in the documentation. However, from a user perspective, a multi-dimensional shift applied on a flattened array makes no sense. The current implementation just applies all shifts in the tuple one after another. 

Wouldn't a much better default for `axis == None` be?:
```diff
a = asanyarray(a)
    if axis is None:
-        return roll(a.ravel(), shift, 0).reshape(a.shape)
+        return roll(a, shift, range(len(shift)))
```

**Comment by mreineck** [comment]
I have been bitten several times by this problem. Since the docstring says about `shift`
```
If a tuple, then `axis` must be a tuple of the same size.
```
 it would be great to enforce this in the code and raise an exception if this condition is not fulfilled, rather than carrying out an operation which the documentation clearly declares invalid.

**Comment by RuRo** [comment]
I was also recently bitten by this. Given the usual `numpy` semantics, I kinda expected
```python
arr = np.array([[1, 0, 0],
                [0, 2, 0],
                [0, 0, 3],
                [4, 0, 5]])
np.roll(arr, np.arange(3), axis=0)
np.roll(arr, np.arange(4), axis=1)
```
To result in
```python
[[1, 0, 3],
 [0, 0, 5],
 [0, 2, 0],
 [4, 0, 0]]
```
and
```python
[[1, 0, 0],
 [0, 0, 2],
 [0, 3, 0],
 [4, 0, 5]]
```

The intuition here would be that `shift` is broadcasted together with `arr` (sans `axis`) and each "matching" 1-D array is rolled a different amount, specified by `shift`. So something along the lines of (but hopefully more efficient):

```python
Ni, M, Nk = a.shape[:axis], a.shape[axis], a.shape[axis+1:]
assert shift.shape == (Ni + Nk)
out = np.empty_like(a)

for ii in np.ndindex(Ni):
    for kk in np.ndindex(Nk):
        match = ii + np.s_[:,] + kk
        out[match] = np.roll(a[match], shift[ii + kk])
```

**Comment by seberg** [comment]
Sounds like we should consider this a bug (although when fixing I would probably just go with a `VisibleDeprecationWarning` just in case).
